Code-level metrics anomaly detection (#14149)
newrelic/docs-website: src/content/docs/codestream/how-use-codestream/performance-monitoring.mdx

  src={codestreamObservabilitySection}
/>

> When hovering over a service name in the tree, click on the globe to open the service's summary page on New Relic. Click on any error to view the details, inclding the stack trace, and [start collaborating](#collaborate) on resolution.

Reviewer: ```suggestion
When hovering over a service name in the tree, click on the globe to open the service's summary page in New Relic. Click on any error to view the details, including the stack trace, and [start collaborating](#collaborate) on resolution.
```

</Callout>

## Code-level metrics [#code-level]

Code-level metrics give you detailed insight into how your code is performing at the method level. Metrics are displayed right in the editor above instrumented methods, allowing performance data to influence the work at hand. CodeStream will also identify poorly performing methods in your code, allowing you to proactively address performance issues as they arise.

### In-editor performance data [#clm-codelenses]

Without having to seek it out, performance data is displayed contextually in the editor so that it can inform the work at hand. Say you’re working on a bug fix, and happen to notice that the area of the code you’re working in has an elevated error rate. Maybe you address that issue as part of your bug fix so that it doesn’t escalate into a bigger problem.

<img
  title="A CodeLens displaying error rate and average duration for a method."
  alt="A screenshot CodeLens displaying error rate and average duration for a method."
  src={codestreamCodeLevelMetricsExample}
/>

> CodeStream will show the average duration and error rate, for the last 30 minutes, in a CodeLens above each method instrumented by the New Relic APM agents. Metrics are calculated for the service currently selected in the Observability section of CodeStream. Expand a different service and the metrics will update accordingly.

Reviewer: ```suggestion
CodeStream will show the average duration and error rate, for the last 30 minutes, in a CodeLens above each method instrumented by the New Relic APM agents. Metrics are calculated for the service currently selected in the **Observability** section of CodeStream. Expand a different service and the metrics will update accordingly.
```

Code-level metrics also help make code review more data driven, with CodeLenses being displayed in pull request diffs reviewed in CodeStream.

<img
  title="A pull request diff with code-level metrics."
  alt="A screenshot of a pull request diff with code-level metrics."
  src={codestreamCodeLevelDiffs}
/>

> Click the CodeLens to see charts visualizing each of the metrics. If the repository is associated with multiple New Relic-monitored services, you can easily switch between them. This will change the service upon which the code-level metrics displayed in both the charts and the editor are based, and also change the service selection in the Observability section.

Reviewer: ```suggestion
Click the CodeLens to see charts visualizing each of the metrics. If the repository is associated with multiple New Relic-monitored services, you can easily switch between them. This will change the service upon which the code-level metrics displayed in both the charts and the editor are based, and also change the service selection in the **Observability** section.
```

<img
  title="Code-level metrics details showing charts and errors."
  alt="A screenshot of code-level metrics details showing charts and errors."
  src={codestreamCodeLevelCharts}
/>

Along with a chart of the error rate you’ll also see a list of the actual errors happening in the same timeframe, including the number of occurrences for each. If the error rate is spiking, and you see one particular error is causing the problem, you can click on it to view the stack trace and [start collaborating](#collaborate) on resolution.

### Anomaly detection [#clm-anomalies]

> CodeLenses are great for when you happen to be working in a section of the code exhibiting a performance problem, but performance issues are too important to just stumble upon! CodeStream will automatically call out poorly performing methods in your code in the Observability > Code-Level Metrics section.

Reviewer: ```suggestion
CodeLenses are great for when you happen to be working in a section of the code exhibiting a performance problem, but performance issues are too important to just stumble upon! CodeStream will automatically call out poorly performing methods in your code in the **Observability > Code-Level Metrics** section.
```

<img
  title="Code-level metrics section showing poorly performing methods."
  alt="A screenshot of the code-level metrics section showing poorly performing methods."
  src={codestreamCodeLevelAnomalies}
/>

CodeStream identifies these methods by looking for anomalies in either error rate or average duration. If you’re using New Relic’s [change tracking](/docs/change-tracking/change-tracking-introduction/) to send deployment information, anomalies are calculated by comparing the performance since your last release against the preceding three weeks. Otherwise we use the last seven days. You can control these timeframes, as well as the thresholds for determining whether or not a given method is performing poorly, by hovering over the “Code-Level Metrics” heading and clicking on the gear icon. 

> In addition to the Code-Level Metrics section, CodeStream will also let you know about poorly performing methods via [desktop notifications](/docs/codestream/codestream-integrations/notifications#additionaldesktop) in either VS Code or JetBrains IDEs.

Reviewer: ```suggestion
In addition to the **Code-Level Metrics** section, CodeStream will also let you know about poorly performing methods via [desktop notifications](/docs/codestream/codestream-integrations/notifications#additionaldesktop) in either VS Code or JetBrains IDEs.
```

### Requirements [#clm-requirements]

In-editor CodeLenses are available for all of New Relic’s APM languages. Anomaly detection, however, is currently only available for Java, .NET, Python and Ruby.

To see metrics, your service must meet the requirements listed below, and the service should have collected data in the last 30 minutes. The New Relic agent attempts to automatically collect data for functions on classes that are tied to HTTP requests. In many cases, and with many frameworks that use an MVC framework, these are often methods on a Controller class. 

> Because all requests by the New Relic agent are not collected all the time, it’s possible that low trafficked methods will not see any data. If data is missing for a specific method that you wish to see results for, custom instrumentation can be used to fill any gaps. See guidance for [Java](/docs/apm/agents/java-agent/custom-instrumentation/java-custom-instrumentation), [.NET](/docs/apm/agents/net-agent/custom-instrumentation/introduction-net-custom-instrumentation), [PHP](/docs/apm/agents/php-agent/features/php-custom-instrumentation), [Python](/docs/apm/agents/python-agent/custom-instrumentation/python-custom-instrumentation), [Ruby](/docs/apm/agents/ruby-agent/api-guides/ruby-custom-instrumentation), [Go](/docs/apm/agents/go-agent/instrumentation/instrument-go-transactions) and [Node.js](/docs/apm/agents/nodejs-agent/extend-your-instrumentation/nodejs-custom-instrumentation). Also note that for Node.js, code-level metrics are not captured for anonymous functions.

Reviewer: ```suggestion
Because all requests by the New Relic agent are not collected all the time, it’s possible that low-traffic methods will not see any data. If data is missing for a specific method that you wish to see results for, you can use custom instrumentation to fill any gaps. See guidance for [Java](/docs/apm/agents/java-agent/custom-instrumentation/java-custom-instrumentation), [.NET](/docs/apm/agents/net-agent/custom-instrumentation/introduction-net-custom-instrumentation), [PHP](/docs/apm/agents/php-agent/features/php-custom-instrumentation), [Python](/docs/apm/agents/python-agent/custom-instrumentation/python-custom-instrumentation), [Ruby](/docs/apm/agents/ruby-agent/api-guides/ruby-custom-instrumentation), [Go](/docs/apm/agents/go-agent/instrumentation/instrument-go-transactions) and [Node.js](/docs/apm/agents/nodejs-agent/extend-your-instrumentation/nodejs-custom-instrumentation). Also note that for Node.js, code-level metrics are not captured for anonymous functions.
```

> * [Distributed tracing](/docs/distributed-tracing/enable-configure/language-agents-enable-distributed-tracing) must be enabled. Distributed tracing is on by default in recent versions of all agents, so you only need to worry about this if you're disabled distributed tracing.

Reviewer: ```suggestion
* [Distributed tracing](/docs/distributed-tracing/enable-configure/language-agents-enable-distributed-tracing) must be enabled. Distributed tracing is on by default in recent versions of all agents, so you only need to worry about this if you've disabled distributed tracing.
```

  Select code in your editor to add a comment.
</figcaption>

> Once you've identified the problem you can assign the error, or update the error status from unresolved to resolved or ignored.

Reviewer: ```suggestion
Once you've identified the problem you can assign the error, or update the error status from `unresolved` to `resolved` or `ignored`.
```
